File: direct-mode/clients/android/app/src/main/java/ai/spatius/avatarkit/directmodedemo/realtime/RealtimeClient.kt
package ai.spatius.avatarkit.directmodedemo.realtime

import android.util.Base64
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

/**
 * The realtime scene's link to the backend agent.
 *
 * Direct Mode either way: the client owns the Motion Server connection and drives the
 * avatar itself. The scenes differ only in where the audio comes from —
 *
 *   pre-recorded  a bundled .pcm file  ──────────────────►  controller.send()
 *   realtime      mic ──ws──► agent (ASR/LLM/TTS) ──ws──►  controller.send()
 *
 * — so both end at the same call and the rendering side is untouched.
 *
 * There is no LiveKit SDK here on purpose. The agent runs server-side without a room:
 * `AgentSession` only builds a RoomIO when its audio input and output are unset, and
 * the backend sets both (see servers/python/realtime.py), so its speech comes back
 * over this plain WebSocket as PCM16.
 */
class RealtimeClient(
    private val callbacks: Callbacks,
) {
    interface Callbacks {
        /** A reply started arriving; the bytes are PCM16 for `controller.send`. */
        fun onAudio(pcm: ByteArray)
        /** The agent finished a reply — the empty final send closes the turn. */
        fun onTurnEnd()
        /** The user talked over the reply; drop what has not played yet. */
        fun onInterrupt()
        fun onTranscript(role: String, text: String)
        fun onError(message: String)
        fun onClosed()
    }

    private var socket: WebSocket? = null

    @Volatile
    var isReady: Boolean = false
        private set

    private val http = OkHttpClient.Builder()
        // The agent holds the socket open between turns, so no read timeout.
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .connectTimeout(15, TimeUnit.SECONDS)
        .build()

    /** Connect, and resolve once the agent is up and listening. */
    suspend fun connect(url: String, language: String): Unit = suspendCoroutine { cont ->
        var settled = false

        val listener = object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                webSocket.send(JSONObject().put("type", "start").put("language", language).toString())
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                val msg = runCatching { JSONObject(text) }.getOrNull() ?: return
                when (msg.optString("type")) {
                    "ready" -> {
                        isReady = true
                        if (!settled) { settled = true; cont.resume(Unit) }
                    }
                    "audio" -> {
                        val encoded = msg.optString("audio")
                        if (encoded.isNotEmpty()) {
                            callbacks.onAudio(Base64.decode(encoded, Base64.DEFAULT))
                        }
                    }
                    "turn_end" -> callbacks.onTurnEnd()
                    "interrupt" -> callbacks.onInterrupt()
                    "transcript" -> callbacks.onTranscript(
                        msg.optString("role"),
                        msg.optString("text"),
                    )
                    "error" -> {
                        val message = msg.optString("message").ifEmpty { "Agent error" }
                        callbacks.onError(message)
                        if (!settled) { settled = true; cont.resumeWithException(IllegalStateException(message)) }
                    }
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                isReady = false
                if (!settled) {
                    settled = true
                    cont.resumeWithException(IllegalStateException("Cannot reach the agent at $url"))
                } else {
                    callbacks.onError(t.message ?: "Agent connection failed")
                }
                callbacks.onClosed()
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                isReady = false
                callbacks.onClosed()
            }
        }

        socket = http.newWebSocket(Request.Builder().url(url).build(), listener)
    }

    /** Microphone audio, as PCM16 at the rate the SDK was initialized with. */
    fun pushMicAudio(pcm: ByteArray) {
        val ws = socket ?: return
        if (!isReady) return
        ws.send(
            JSONObject()
                .put("type", "mic_audio")
                .put("audio", Base64.encodeToString(pcm, Base64.NO_WRAP))
                .toString()
        )
    }

    /** Speak a fixed line, for trying the scene without a microphone. */
    fun say(text: String) {
        socket?.send(JSONObject().put("type", "text").put("text", text).toString())
    }

    fun interrupt() {
        socket?.send(JSONObject().put("type", "interrupt").toString())
    }

    fun close() {
        isReady = false
        socket?.close(1000, null)
        socket = null
    }
}

File: direct-mode/clients/ios/AvatarDemo/Services/RealtimeClient.swift
import Foundation

/// The realtime scene's link to the backend agent.
///
/// Direct Mode either way: the client owns the Motion Server connection and drives the
/// avatar itself. The scenes differ only in where the audio comes from —
///
///   pre-recorded  a bundled .pcm file  ──────────────────►  controller.send()
///   realtime      mic ──ws──► agent (ASR/LLM/TTS) ──ws──►  controller.send()
///
/// — so both end at the same call and the rendering side is untouched.
///
/// There is no LiveKit SDK here on purpose. The agent runs server-side without a room:
/// `AgentSession` only builds a RoomIO when its audio input and output are unset, and
/// the backend sets both (see servers/python/realtime.py), so its speech comes back
/// over this plain WebSocket as PCM16.
final class RealtimeClient: NSObject {

    struct Callbacks {
        /// A reply started arriving; the bytes are PCM16 for `controller.send`.
        var onAudio: (Data) -> Void = { _ in }
        /// The agent finished a reply — the empty final send closes the turn.
        var onTurnEnd: () -> Void = {}
        /// The user talked over the reply; drop what has not played yet.
        var onInterrupt: () -> Void = {}
        var onTranscript: (String, String) -> Void = { _, _ in }
        var onError: (String) -> Void = { _ in }
        var onClosed: () -> Void = {}
    }

    private var task: URLSessionWebSocketTask?
    private let callbacks: Callbacks
    private var readyContinuation: CheckedContinuation<Void, Error>?

    private(set) var isReady = false

    init(callbacks: Callbacks) {
        self.callbacks = callbacks
        super.init()
    }

    /// Connect, and resolve once the agent is up and listening.
    func connect(url: String, language: String) async throws {
        guard let endpoint = URL(string: url) else {
            throw NSError(domain: "RealtimeClient", code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Cannot reach the agent at \(url)",
            ])
        }
        let session = URLSession(configuration: .default)
        let socket = session.webSocketTask(with: endpoint)
        task = socket
        socket.resume()
        receive()

        let start = try JSONSerialization.data(
            withJSONObject: ["type": "start", "language": language]
        )
        try await socket.send(.string(String(decoding: start, as: UTF8.self)))

        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            readyContinuation = cont
        }
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure(let error):
                self.isReady = false
                if let cont = self.readyContinuation {
                    self.readyContinuation = nil
                    cont.resume(throwing: error)
                } else {
                    self.callbacks.onError(error.localizedDescription)
                }
                self.callbacks.onClosed()
            case .success(let message):
                if case .string(let text) = message {
                    self.handle(text)
                }
                // One receive only ever yields one message, so the next has to be
                // asked for or the socket goes quiet after the first.
                self.receive()
            }
        }
    }

    private func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let msg = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = msg["type"] as? String else { return }

        switch type {
        case "ready":
            isReady = true
            if let cont = readyContinuation {
                readyContinuation = nil
                cont.resume()
            }
        case "audio":
            if let encoded = msg["audio"] as? String,
               let pcm = Data(base64Encoded: encoded), !pcm.isEmpty {
                callbacks.onAudio(pcm)
            }
        case "turn_end":
            callbacks.onTurnEnd()
        case "interrupt":
            callbacks.onInterrupt()
        case "transcript":
            callbacks.onTranscript(
                msg["role"] as? String ?? "",
                msg["text"] as? String ?? ""
            )
        case "error":
            let message = msg["message"] as? String ?? "Agent error"
            callbacks.onError(message)
            if let cont = readyContinuation {
                readyContinuation = nil
                cont.resume(throwing: NSError(domain: "RealtimeClient", code: 1, userInfo: [
                    NSLocalizedDescriptionKey: message,
                ]))
            }
        default:
            break
        }
    }

    /// Microphone audio, as PCM16 at the rate the SDK was initialized with.
    func pushMicAudio(_ pcm: Data) {
        guard isReady, let task else { return }
        let payload: [String: Any] = [
            "type": "mic_audio",
            "audio": pcm.base64EncodedString(),
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload) else { return }
        task.send(.string(String(decoding: data, as: UTF8.self))) { _ in }
    }

    /// Speak a fixed line, for trying the scene without a microphone.
    func say(_ text: String) {
        send(["type": "text", "text": text])
    }

    func interrupt() {
        send(["type": "interrupt"])
    }

    func close() {
        isReady = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    private func send(_ payload: [String: Any]) {
        guard let task, let data = try? JSONSerialization.data(withJSONObject: payload) else { return }
        task.send(.string(String(decoding: data, as: UTF8.self))) { _ in }
    }
}
